File: csharp/FountainTools/Writer.cs
using System.Text.RegularExpressions;

namespace Fountain;

public class FountainWriter
{
    public bool PrettyPrint { get; set; }
    private string? _lastChar;

    public FountainWriter()
    {
        PrettyPrint = true;
        _lastChar = null;
    }

    // Expects FountainScript and returns a UTF-8 formatted string
    public string Write(FountainScript script)
    {
        var lines = new List<string>();

        // Write headers
        if (script.Headers.Count > 0)
        {
            foreach (var header in script.Headers)
            {
                lines.Add(WriteElement(header));
            }

            lines.Add(""); // Add a blank line after headers
        }

        // Write elements
        FountainElement? lastElem = null;

        foreach (var element in script.Elements)
        {
            // Determine padding
            bool padBefore = false;
            if (element.Type == Element.CHARACTER || 
                element.Type == Element.TRANSITION || 
                element.Type == Element.HEADING)
            {
                padBefore = true;
            }
            else if (element.Type == Element.ACTION)
            {
                padBefore = lastElem == null || lastElem.Type != Element.ACTION;
            }

            if (padBefore)
                lines.Add("");

            lines.Add(WriteElement(element));
            lastElem = element;
        }

        // Join lines into a single text
        string text = string.Join("\n", lines);

        // Replace notes
        var regexNotes = new Regex(@"\[\[(\d+)\]\]");
        text = regexNotes.Replace(text, match =>
        {
            int num = int.Parse(match.Groups[1].Value);
            return $"[[{script.Notes[num].Text}]]";
        });

        // Replace boneyards
        var regexBoneyards = new Regex(@"/\*(\d+)\*/");
        text = regexBoneyards.Replace(text, match =>
        {
            int num = int.Parse(match.Groups[1].Value);
            return $"/*{script.Boneyards[num].Text}*/";
        });

        return text;
    }

    private string WriteElement(FountainElement elem)
    {
        switch (elem.Type)
        {
            case Element.CHARACTER:
                return WriteCharacter((FountainCharacter)elem);

            case Element.DIALOGUE:
                return WriteDialogue((FountainDialogue)elem);

            case Element.PARENTHESIS:
                return WriteParenthesis((FountainParenthesis)elem);

            case Element.ACTION:
                return WriteAction((FountainAction)elem);

            case Element.LYRIC:
                return $"~ {elem.TextRaw}";

            case Element.SYNOPSIS:
                return $"= {elem.TextRaw}";

            case Element.TITLEENTRY:
                return $"{((FountainTitleEntry)elem).Key}: {elem.TextRaw}";

            case Element.HEADING:
                return WriteHeading((FountainHeading)elem);

            case Element.TRANSITION:
                return WriteTransition((FountainTransition)elem);

            case Element.PAGEBREAK:
                return "===";

            case Element.SECTION:
                return $"{new string('#', ((FountainSection)elem).Level)} {elem.TextRaw}";

            default:
                _lastChar = null;
                return string.Empty;
        }
    }

    private string WriteCharacter(FountainCharacter elem)
    {
        string pad = PrettyPrint ? new string('\t', 3) : string.Empty;
        string charText = elem.Name;

        if (elem.IsDualDialogue)
            charText += " ^";
        if (!string.IsNullOrEmpty(elem.Extension))
            charText += $" ({elem.Extension})";
        if (elem.Forced)
            charText = "@" + charText;
        if (_lastChar == elem.Name)
            charText += " (CONT'D)";

        _lastChar = elem.Name;
        return $"{pad}{charText}";
    }

    private string WriteDialogue(FountainDialogue elem)
    {
        string output = elem.TextRaw;

        // Ensure blank lines in dialogue have at least a space
        output = string.Join("\n", output.Split('\n')
            .Select(line => string.IsNullOrWhiteSpace(line) ? " " : line));

        // Add tab for pretty printing
        if (PrettyPrint)
        {
            output = string.Join("\n", output.Split('\n')
                .Select(line => $"\t{line}"));
        }

        return output;
    }

    private string WriteParenthesis(FountainParenthesis elem)
    {
        string pad = PrettyPrint ? new string('\t', 2) : string.Empty;
        return $"{pad}({elem.TextRaw})";
    }

    private string WriteAction(FountainAction elem)
    {
        if (elem.Forced)
            return $"!{elem.TextRaw}";
        if (elem.Centered)
            return $">{elem.TextRaw}<";
        return elem.TextRaw;
    }

    private string WriteHeading(FountainHeading elem)
    {
        string sceneNum = !string.IsNullOrEmpty(elem.SceneNum) ? $" #{elem.SceneNum}#" : string.Empty;
        if (elem.Forced)
            return $"\n.{elem.TextRaw}{sceneNum}";
        return $"\n{elem.TextRaw}{sceneNum}";
    }

    private string WriteTransition(FountainTransition elem)
    {
        string pad = PrettyPrint ? new string('\t', 4) : string.Empty;
        if (elem.Forced)
            return $">{elem.TextRaw}";
        return $"{pad}{elem.TextRaw}";
    }
}

File: python/src/fountain_tools/writer.py
import re
from .fountain import Element 


class FountainWriter:
    def __init__(self):
        self.pretty_print = True
        self._last_char = None

    # Expects a FountainScript-like object
    # Returns a UTF-8 string
    def write(self, script):
        lines = []

        # Process headers
        if len(script.headers) > 0:
            for header in script.headers:
                lines.append(self._write_elem(header))
            lines.append("")

        last_elem = None

        # Process elements
        for element in script.elements:
            pad_before = False

            if element.type in [Element.CHARACTER, Element.TRANSITION, Element.HEADING]:
                pad_before = True
            elif element.type == Element.ACTION:
                pad_before = not last_elem or last_elem.type != Element.ACTION

            if pad_before:
                lines.append("")

            lines.append(self._write_elem(element))
            last_elem = element

        # Join lines
        text = "\n".join(lines)

        # Replace notes
        regex_notes = r"\[\[(\d+)\]\]"
        text = re.sub(
            regex_notes,
            lambda match: f"[[{script.notes[int(match.group(1))].text}]]",
            text,
        )

        # Replace boneyards
        regex_boneyards = r"/\*(\d+)\*/"
        text = re.sub(
            regex_boneyards,
            lambda match: f"/*{script.boneyards[int(match.group(1))].text}*/",
            text,
        )

        return text

    def _write_elem(self, elem):
        elem_type = elem.type

        if elem_type == Element.CHARACTER:
            pad = "\t" * 3 if self.pretty_print else ""
            char = elem.name

            if elem.is_dual_dialogue:
                char += " ^"
            if elem.extension:
                char += f" ({elem.extension})"
            if elem.forced:
                char = "@" + char
            if self._last_char == elem.name:
                char += " (CONT'D)"
            self._last_char = elem.name
            return f"{pad}{char}"

        if elem_type == Element.DIALOGUE:
            output = elem._text

            # Ensure blank lines in dialogue have at least a space
            output = "\n".join(line if line.strip() else " " for line in output.split("\n"))

            if self.pretty_print:
                output = "\n".join(f"\t{line}" for line in output.split("\n"))

            return output

        if elem_type == Element.PARENTHESIS:
            pad = "\t" * 2 if self.pretty_print else ""
            return f"{pad}({elem._text})"

        if elem_type == Element.ACTION:
            if elem.forced:
                return f"!{elem._text}"
            if elem.centered:
                return f">{elem._text}<"
            return elem._text

        if elem_type == Element.LYRIC:
            return f"~ {elem._text}"
        
        if elem_type == Element.SYNOPSIS:
            return f"= {elem._text}"

        self._last_char = None

        if elem_type == Element.TITLEENTRY:
            return f"{elem.key}: {elem._text}"

        if elem_type == Element.HEADING:
            scene_num = f" #{elem.scene_num}#" if elem.scene_num else ""
            if elem.forced:
                return f"\n.{elem._text}{scene_num}"
            return f"\n{elem._text}{scene_num}"

        if elem_type == Element.TRANSITION:
            pad = "\t" * 4 if self.pretty_print else ""
            if elem.forced:
                return f">{elem._text}"
            return f"{pad}{elem._text}"

        if elem_type == Element.PAGEBREAK:
            return "==="

        if elem_type == Element.SECTION:
            return f"{'#' * elem.level} {elem.text}"

        return ""